### DISCUSSION TITLE
Problems with Projection onto the SOS Cone

### DISCUSSION THREAD

**Discussion by JOlucak** [iFR-ACSO/casos#83] [Q&A]
We are interested to solve the projection onto the SOS (or any other cone) to obtain the constraint violation. In CaSoS this can be easily done for some constraint e.g.

    s    = casos.PS.sym('q',grambasis( constraint ) );

       % projection error
        e = s - constraint;
        
        
        % define SOS problem 
        %   min ||s-p||^2  s.t. s is SOS
        proj_sos = struct('x',s,...
                                   'f',dot(e,e);
        
        opts.Kx            = struct('sos', length(s));
        opts.error_on_fail = 1;
        obj.projConPara    =  casos.sossol('S','mosek',proj_sos,opts);

However, if Mosek is used as solver, it frequently runs into numerical issues. 
One  could "ignore" this by simply setting          opts.error_on_fail = 0 what is obviously not recommended. 
Also re-writting the problem with linear decision variables and additional SOS constraint might not help here.

This issue needs furhter investigation.  

Points to be clarified:

- The computation of the projection should be always possible? 
- How trustworthy is the mosek "result" in case of numerical problems? Obviously, mosek has problem in solving the problem but the question is where these problems come from.
- Which alternatives do we have?

**Comment by tcunis** [comment]
> The computation of the projection should be always possible?

Yes, the projection onto a convex set (cone) is a strictly convex problem and hence always has a (unique) solution.

> How trustworthy is the mosek "result" in case of numerical problems?

You should check first if the solution Mosek returns for `x` is feasible (i.e., sum-of-squares). If it is not, the result is not to be trusted. If it is, you could check whether `constraint` is sum-of-squares. 

> Which alternatives do we have?

Possible alternatives:

1. Other solver: SCS supports quadratic objective functions natively, thus avoiding the epigraph formulation.
2. Other objective: Project using L1 or Linf norm, again avoiding the quadratic cost function and its epigraph formulation.
3. Scaling: It is possible the problem is illconditioned, possibly due to scaling of the polynomial coefficients. You could try using `pnorm2` for the objective function which defines an L2 norm on the polynomials via an integral.
